File: testdata/e2e.py
import json, subprocess, os

def frame(m):
    b = json.dumps(m).encode()
    return b"Content-Length: %d\r\n\r\n" % len(b) + b

def read_msg(f):
    while True:
        line = f.readline()
        if not line: return None
        if line.startswith(b"Content-Length:"):
            length = int(line.split(b":")[1])
        elif line in (b"\r\n", b"\n"):
            break
    return json.loads(f.read(length))

doc = "package x\n\nfunc h() error {\n\terr := work()\n\tif err !="
e2 = dict(os.environ)
e2.update({"Q4COMPLETE_MODEL": "bin/model.bin", "Q4COMPLETE_JOURNAL": "testdata/journal_test.jsonl"})
p = subprocess.Popen(["./bin/q4complete", "serve"], stdin=subprocess.PIPE, stdout=subprocess.PIPE, env=e2)
p.stdin.write(frame({"jsonrpc":"2.0","id":1,"method":"initialize","params":{}}))
p.stdin.write(frame({"jsonrpc":"2.0","method":"textDocument/didOpen","params":{"textDocument":{"uri":"file:///x.go","languageId":"go","version":1,"text":doc}}}))
p.stdin.write(frame({"jsonrpc":"2.0","id":2,"method":"textDocument/inlineCompletion","params":{"textDocument":{"uri":"file:///x.go"},"position":{"line":4,"character":10}}}))
p.stdin.write(frame({"jsonrpc":"2.0","id":3,"method":"q4/learn","params":{"text":"\tif err != nil {","uri":"file:///x.go","line":4}}))
p.stdin.write(frame({"jsonrpc":"2.0","id":4,"method":"textDocument/inlineCompletion","params":{"textDocument":{"uri":"file:///x.go"},"position":{"line":4,"character":10}}}))
p.stdin.write(frame({"jsonrpc":"2.0","id":6,"method":"q4/reject","params":{"uri":"file:///x.go","line":4}}))
p.stdin.write(frame({"jsonrpc":"2.0","id":7,"method":"q4/status","params":{}}))
p.stdin.write(frame({"jsonrpc":"2.0","id":5,"method":"shutdown","params":{}}))
p.stdin.write(frame({"jsonrpc":"2.0","method":"exit","params":{}}))
p.stdin.flush()
while True:
    m = read_msg(p.stdout)
    if m is None: break
    print(json.dumps(m)[:400])
p.wait(timeout=30)
print("server exit:", p.returncode)

File: testdata/e2e.mjs
import { spawn } from "node:child_process";

const srv = spawn("bin/q4complete", ["serve"], { env: { ...process.env, Q4COMPLETE_MODEL: "bin/model.bin" } });
let buf = Buffer.alloc(0);
const pending = new Map();
let id = 0;

srv.stdout.on("data", (d) => {
  buf = Buffer.concat([buf, d]);
  for (;;) {
    const m = buf.toString("latin1").match(/Content-Length: (\d+)\r\n\r\n/);
    if (!m) break;
    const len = +m[1];
    const start = m.index + m[0].length;
    if (buf.length < start + len) break;
    const body = JSON.parse(buf.subarray(start, start + len).toString());
    buf = buf.subarray(start + len);
    if (body.id !== undefined && pending.has(body.id)) {
      pending.get(body.id)(body);
      pending.delete(body.id);
    }
  }
});
srv.stderr.on("data", (d) => process.stderr.write("[srv] " + d));

function send(method, params, isReq = true) {
  const msg = { jsonrpc: "2.0", method, params };
  if (isReq) msg.id = ++id;
  const body = JSON.stringify(msg);
  srv.stdin.write(`Content-Length: ${body.length}\r\n\r\n${body}`);
  if (!isReq) return Promise.resolve(null);
  return new Promise((res) => pending.set(id, res));
}

const doc = "package main\n\nfunc check() error {\n\tif err := doThing(); e";

await send("initialize", { processId: process.pid, capabilities: {}, rootUri: null });
await send("initialized", {}, false);
await send("textDocument/didOpen", { textDocument: { uri: "file:///t.go", languageId: "go", version: 1, text: doc } }, false);
await new Promise((r) => setTimeout(r, 50));
const res = await send("q4/inlineCompletion", {
  textDocument: { uri: "file:///t.go" },
  position: { line: 4, character: 23 },
});
console.log("RESULT", JSON.stringify(res.result));
const res2 = await send("q4/status", {});
console.log("STATUS", JSON.stringify(res2.result));
await send("shutdown", {});
await send("exit", {}, false);
srv.kill();
process.exit(0);
